### PR TITLE
Update restricted imports for Calypso.

### PR DESCRIPTION
**Pull request by sgomes**
This is a follow-up to [a comment in another PR](https://github.com/Automattic/wp-calypso/pull/35783#discussion_r318521481).

It adds two changes to the ESLint config:

- Add `gridicons` restricted import. We want to migrate away from the "naked" import, and use `components/gridicon` explicitly, instead of relying on the webpack alias forever.
- Rewrite `wpcalypso/import-no-redux-combine-reducers` as a restricted import.

~~Note that this does not delete the `wpcalypso/import-no-redux-combine-reducers` rule code, to avoid breaking compatibility for consumers of the `eslint-plugin-wpcalypso` package.~~ After some feedback from @sirreal, I removed the rule.

#### Testing instructions

* Ensure that any file with `import Gridicon from 'gridicons';` is marked as having a lint error.
* Add `import { combineReducers } from 'redux';` to any file, and ensure that it's marked as having a lint error.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -106,9 +106,40 @@ module.exports = {
...
+			{
+				paths: [
+					// Error if any module depends on the data-observe mixin, which is deprecated.
+					'lib/mixins/data-observe',
+					// Prevent naked import of gridicons module. Use 'components/gridicon' instead.
+					{
+						name: 'gridicons',
+						message: "Please use 'components/gridicon' instead.",
+					},
+				],
```
Comment (sgomes) [inline .eslintrc.js]: Note that we can't check for Redux `combineReducers` in CommonJS, as there's no concept of a non-default import there.

**Review by sirreal — COMMENTED** [review]
I think we can get rid of all references to the obsolete rule, I don't believe it serves any purpose:

https://github.com/Automattic/wp-calypso/blob/274236f6ebec72ea56e056da44f5bab320670c53/packages/eslint-plugin-wpcalypso/lib/rules/import-no-redux-combine-reducers.js#L9-L13
https://github.com/Automattic/wp-calypso/blob/e5ff4bb5fd98726827b35ae261908bc0afd43b6e/packages/eslint-plugin-wpcalypso/lib/rules/index.js#L12
https://github.com/Automattic/wp-calypso/blob/e5ff4bb5fd98726827b35ae261908bc0afd43b6e/packages/eslint-plugin-wpcalypso/lib/rules/test/import-no-redux-combine-reducers.js
https://github.com/Automattic/wp-calypso/blob/e5ff4bb5fd98726827b35ae261908bc0afd43b6e/packages/eslint-plugin-wpcalypso/lib/rules/index.js#L12

**Comment by sirreal** [comment]
> Note that this does not delete the wpcalypso/import-no-redux-combine-reducers rule code, to avoid breaking compatibility for consumers of the eslint-plugin-wpcalypso package.

We're due to update the package soon anyways, we can add a changelog entry as well so it's clear this is a major update.

One or both of the eslint packages will be merged into the other so there's a single package with rules and config. That wasn't possible when the packages were created.

**Comment by sgomes** [comment]
> We're due to update the package soon anyways, we can add a changelog entry as well so it's clear this is a major update.
> 
> One or both of the eslint packages will be merged into the other so there's a single package with rules and config. That wasn't possible when the packages were created.

Thank you! I removed the rule altogether, and updated everything accordingly (I think).

**Review by sirreal — APPROVED** [review]
👍 This seems to be working nicely, thanks!

It's too bad we can't implement eslint `fix` for the gridicons rule. Maybe a codemod or simple `sed` is in order 🙂

**Comment by sgomes** [comment]
> Maybe a codemod or simple `sed` is in order 🙂

I shudder to think of all the collisions that would entail, both in the codemod PR (while it's being reviewed), and all of the open PRs in Calypso after it gets merged 😅 Still, it's an option.

**Pull request merged**